mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java

```diff
@@ -283,14 +278,51 @@ public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws Mailb
     @Override
     public Iterator<UpdatedFlags> updateFlags(Mailbox mailbox, FlagsUpdateCalculator flagUpdateCalculator, MessageRange set) throws MailboxException {
         CassandraId mailboxId = (CassandraId) mailbox.getMailboxId();
-        return messageIdDAO.retrieveMessages(mailboxId, set)
-                .join()
-                .flatMap(message -> updateFlagsOnMessage(mailbox, flagUpdateCalculator, message))
-                .map((UpdatedFlags updatedFlags) -> indexTableHandler.updateIndexOnFlagsUpdate(mailboxId, updatedFlags)
-                    .thenApply(voidValue -> updatedFlags))
-                .map(CompletableFuture::join)
-                .collect(Collectors.toList()) // This collect is here as we need to consume all the stream before returning result
-                .iterator();
+
+        return runUpdate(mailboxId, set, flagUpdateCalculator).iterator();
+    }
+
+    private List<UpdatedFlags> runUpdate(CassandraId mailboxId, MessageRange set, FlagsUpdateCalculator flagsUpdateCalculator) throws MailboxException {
+        Stream<ComposedMessageIdWithMetaData> toBeUpdated = messageIdDAO.retrieveMessages(mailboxId, set).join();
+
+        FlagsUpdateStageResult globalResult = runUpdateStage(mailboxId, toBeUpdated, flagsUpdateCalculator);
+
+        int retryCount = 0;
+
+        while (retryCount < maxRetries && !globalResult.getFailed().isEmpty()) {
+            retryCount++;
+            FlagsUpdateStageResult stageResult = runUpdateStage(mailboxId,
+                FluentFutureStream.of(
+                    globalResult.getFailed().stream()
+                        .map(uid -> messageIdDAO.retrieve(mailboxId, uid)))
+                    .flatMap(OptionalConverter::toStream)
+                    .completableFuture().join(),
+                flagsUpdateCalculator);
+
+            globalResult = globalResult.keepSuccess().merge(stageResult);
+        }
+
+        LOGGER.error("Can not update following UIDs {} for mailbox {}", globalResult.getFailed(), mailboxId.asUuid());
+
+        return globalResult.getSucceeded();
+    }
+
+    private FlagsUpdateStageResult runUpdateStage(CassandraId mailboxId, Stream<ComposedMessageIdWithMetaData> toBeUpdated, FlagsUpdateCalculator flagsUpdateCalculator) {
+        Long newModSeq = modSeqProvider.nextModSeq(mailboxId).join().orElseThrow(() -> new RuntimeException("ModSeq generation failed"));
+
+        FlagsUpdateStageResult result = toBeUpdated
+            .map(oldMetadata -> tryFlagsUpdate(flagsUpdateCalculator,
+                newModSeq,
+                oldMetadata))
+            .reduce(FlagsUpdateStageResult::merge)
+            .orElse(none());
+
+        result.getSucceeded().stream()
+            .map((UpdatedFlags updatedFlags) -> indexTableHandler.updateIndexOnFlagsUpdate(mailboxId, updatedFlags)
```

Comment:
Well, if a meteor falls on our servers, as we do not have transaction semantic across tables, this is not something that can be solved easily.